### DISCUSSION TITLE
ilike behavior has changed in the 2.0.x

### DISCUSSION THREAD

**Discussion by alexted** [sqlalchemy/sqlalchemy#9508] [Usage Questions] [answered]
my model is:
```python
class Country(Base):
    __tablename__ = "country"

    id = Column(Integer, primary_key=True)  # noqa: A003
    country_code = Column(String(2), nullable=False, unique=True, index=True)
    country_full = Column(String(100), nullable=False)
    tier = Column(SmallInteger, nullable=False)
    aliases = Column(ARRAY(Text), nullable=False)

    def __repr__(self):
        return f"Country(id={self.id}, country_code={self.country_code}, country_full={self.country_full})"
```

my query is:
```python
stmt = (
    select(Country)
    .where(
        or_(
            Country.country_code.ilike(f"%{pattern}%"),
            Country.country_full.ilike(f"%{pattern}%"),
            func.array_to_string(Country.aliases, ",").ilike(f"%{pattern}%"),
        ),
    )
    .order_by(Country.country_full)
)

```
in 1.14.47 it produces the expression:
```sql
SELECT * 
FROM country 
WHERE country.country_code ILIKE '%tai%' 
OR country.country_full ILIKE '%tai%' 
OR array_to_string(country.aliases, ',') ILIKE '%tai%'
ORDER BY country.country_full
;
```
in 2.0.7 it produces the expression:
```sql
SELECT * 
FROM country 
WHERE country.country_code ILIKE '%tai%'::VARCHAR(2) 
OR country.country_full ILIKE '%tai%'::VARCHAR(100) 
OR array_to_string(country.aliases, ','::VARCHAR) ILIKE '%tai%'::VARCHAR
ORDER BY country.country_full
;
```

which gave me a completely different sampling result:
in 1.14.x
![image](https://user-images.githubusercontent.com/33978094/226160185-c2e5ec57-17e3-4168-bffa-e63af53ae1cf.png)
vs
in 2.0.x
![image](https://user-images.githubusercontent.com/33978094/226160217-c60b804f-9f89-4bfe-be4a-db170baf0a13.png)
I use PostgreSQL 11.4
Could you tell me what the problem is and how to fix it?

**Comment by CaselIT** [comment]
Hi,

I think the issue is in the country code part. Use a bindparam specifying the type, so that it doesn't inherited the type of the left side.

@zzzeek strange that the types are set like this. IIRC they should not https://github.com/sqlalchemy/sqlalchemy/issues/9313

**Reply by zzzeek** [reply]
nothing has changed here since 1.4 except that we have to cast parameters for asyncpg and psycopg.

the BindParam receives the incoming value, a fixed datatype that was passed in if any, and the compared to datatype.  if it has no fixed datatype and only a compared to datatype, it uses [coerce_compared_value](https://github.com/sqlalchemy/sqlalchemy/blob/b56494ba64f858a60a3314fd549ced4512a5fe19/lib/sqlalchemy/sql/type_api.py#L1041) to get the resulting type.   

hence this is a bug and should be fixed as:

```diff
diff --git a/lib/sqlalchemy/sql/sqltypes.py b/lib/sqlalchemy/sql/sqltypes.py
index 4583948704..ae90f74c95 100644
--- a/lib/sqlalchemy/sql/sqltypes.py
+++ b/lib/sqlalchemy/sql/sqltypes.py
@@ -257,6 +257,14 @@ class String(Concatenable, TypeEngine[str]):
     def get_dbapi_type(self, dbapi):
         return dbapi.STRING
 
+    def coerce_compared_value(
+        self, op: Optional[OperatorType], value: Any
+    ) -> TypeEngine[Any]:
+        sup = super().coerce_compared_value(op, value)
+        if sup is self:
+            return self._resolve_for_literal(value)
+        else:
+            return sup
 
 class Text(String):

```

**Reply by zzzeek** [reply, marked as answer]
this is #9511.  this is a fairly critical regression since it wrecks queries without raising an error
